### DISCUSSION TITLE
Add metadata to individuals

### DISCUSSION THREAD

**Discussion by andrei-barysenka** [tskit-dev/msprime#2021] [Q&A] [answered]
How can I add metadata info to individuals in my tree sequence ? I do something like this:

ts1=msprime.sim_ancestry( ... )
for I in ts1.individuals():
        I.metadata=b"META"
for I in ts1.individuals():
        print(I)

but metadata is still empty

**Comment by jeromekelleher** [comment, marked as answer]
Hi @andrei-barysenka :wave:

The [TreeSequence](https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence) object is read-only, and any modifications that you make to the Python objects you get back will either fail, or be transitory (as you have here with the Individual objects). If you want the changes to "stick" you need to work with the underlying [TableCollection](https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TableCollection). For your example:

```python
import msprime

ts = msprime.sim_ancestry(3, random_seed=2)

tables = ts.dump_tables()
tables.individuals.clear()
for individual in ts.individuals():
    individual.metadata = b"META"
    tables.individuals.append(individual)

print(tables.individuals)
ts2 = tables.tree_sequence()
for i in ts2.individuals():
    print(i)
```

However, you probably want to use a schema if you're adding metadata. This is the simplest way:

```python
import tskit

tables = ts.dump_tables()
tables.individuals.clear()
tables.individuals.metadata_schema = tskit.MetadataSchema.permissive_json()
for individual in ts.individuals():
    individual.metadata = {"key": "value"}
    tables.individuals.append(individual)

print(tables.individuals)
ts2 = tables.tree_sequence()
for i in ts2.individuals():
    print(i)

```
You might find these tutorials useful:

- https://tskit.dev/tutorials/tables_and_editing.html
- https://tskit.dev/tutorials/metadata.html
